File: 86-partition-list/go/main.go
package main

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func partition(head *ListNode, x int) *ListNode {
	smallHead := &ListNode{Val: 0, Next: nil}
	bigHead := &ListNode{Val: 0, Next: nil}
	s := smallHead
	b := bigHead
	for head != nil {
		if head.Val < x {
			s.Next = head
			s = s.Next
		} else {
			b.Next = head
			b = b.Next
		}
		head = head.Next
	}

	s.Next = bigHead.Next
	b.Next = nil
	return smallHead.Next
}

File: 86-partition-list/cpp/main.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
 public:
  ListNode* partition(ListNode* head, int x) {
    ListNode* less = new ListNode();
    ListNode* more = new ListNode();
    ListNode* lessI = less;
    ListNode* moreI = more;

    while (head != NULL) {
      if (head->val < x) {
        lessI->next = head;
        lessI = lessI->next;
      } else {
        moreI->next = head;
        moreI = moreI->next;
      }
      head = head->next;
    }

    moreI->next = NULL;
    lessI->next = more->next;
    return less->next;
  }
};
